DOC-3850 Vault and Softlocking key concept (#586)
corda/corda-docs-portal: content/en/platform/corda/4.8/open-source/key-concepts-vault.md

     * input refs are consumed as part of cash spending.
     */
    fun softLockRelease(lockId: UUID, stateRefs: NonEmptySet<StateRef>? = null)

```

[VaultService.kt](https://github.com/corda/corda/blob/release/os/4.8/core/src/main/kotlin/net/corda/core/node/services/VaultService.kt)

### Querying the vault with `SoftLockingCondition`

By default, vault queries always include locked states in its result sets. Custom filterable criteria can be specified using the `SoftLockingCondition` attribute of `VaultQueryCriteria`:

```kotlin
    @CordaSerializable
    data class SoftLockingCondition(val type: SoftLockingType, val lockIds: List<UUID> = emptyList())

    @CordaSerializable
    enum class SoftLockingType {
        UNLOCKED_ONLY,  // only unlocked states
        LOCKED_ONLY,    // only soft locked states
        SPECIFIED,      // only those soft locked states specified by lock id(s)
        UNLOCKED_AND_SPECIFIED   // all unlocked states plus those soft locked states specified by lock id(s)
    }

```

[QueryCriteria.kt](https://github.com/corda/corda/blob/release/os/4.8/core/src/main/kotlin/net/corda/core/node/services/vault/QueryCriteria.kt)

### Explicit Usage

> Soft locks are associated with transactions, and typically within the lifecycle of a flow. Specifically, every time a
> flow is started a soft lock identifier is associated with that flow for its duration (and released upon it’s natural
> termination or in the event of an exception). The `VaultSoftLockManager` is responsible within the Node for

Reviewer: ```suggestion
Soft locks are associated with transactions, typically within the lifecycle of a flow. Every time a
flow starts, a soft lock identifier is associated with that flow for its duration and released upon its natural
termination or in the event of an exception. The `VaultSoftLockManager` is responsible within the node for
```
Could avoid the parentheses here by splitting the sentence.

automatically managing this soft lock registration and release process for flows. The `TransactionBuilder` class has a
new `lockId` field for the purpose of tracking lockable states. By default, it is automatically set to a random
`UUID` (outside of a flow) or to a flow’s unique ID (within a flow).

Upon building a new transaction to perform some action for a set of states on a contract, a developer must explicitly

> register any states they may wish to hold until that transaction is committed to the ledger. These states will be effectively ‘soft
> locked’ (not usable by any other transaction) until the developer explicitly releases these or the flow terminates or errors

Reviewer: ```suggestion
register any states they may wish to hold until that transaction is committed to the ledger. These states are effectively ‘soft
locked’ (not usable by any other transaction) until the developer explicitly releases these or the flow terminates or errors
```

(at which point they are automatically released).

## An example of soft locking in action

A prime example where *soft locking* is automatically enabled is within the process of issuance and transfer of fungible
state, like cash.

> For example, Bank of Corda - an issuer of fungible assets - wants to transfer newly issued assets immediately
> to Big Corp, the issuance requester. This issuance and transfer operation must be *atomic*, such that another

Reviewer: I think it would be too easy for people to think Bank of Corda is a real thing 😅 Suggest sticking with poor old Alice and Bob here.

```suggestion
For example, Alice—an issuer of fungible assets—wants to transfer newly issued assets immediately
to Bob, the issuance requester. This issuance and transfer operation must be *atomic*, such that another
```

> flow, or instance of the same flow, does not step in and unintentionally spend the states issued by Bank of Corda

Reviewer: ```suggestion
flow, or instance of the same flow, does not step in and unintentionally spend the states issued by Alice
```
